### PR TITLE
fix: nix: add platform dependent flag to lib target

### LINKED ISSUE #3810: Shared library fails to build on macOS with Nix
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

The lean4 nix flake fails to build on macOS:
```
error: builder for '/nix/store/v9bw3gi6w3zxxzp1xpcjvafcqpvnq0il-libInit_shared.drv' failed with exit code 1;
       last 3 log lines:
       > ld: warning: directory not found for option '-L/nix/store/7bgg3pmrhjm59sl1xfrd0b9ry94cd9l3-lean-bin-tools/lib/lean'
       > ld: unknown option: -Bsymbolic
       > clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
       For full logs, run 'nix-store -l /nix/store/v9bw3gi6w3zxxzp1xpcjvafcqpvnq0il-libInit_shared.drv'.
error: 1 dependencies of derivation '/nix/store/c7sp7q2fns3275i9b2da43dhrsr77qp5-lean.drv' failed to build
```

### Context

This occurs when using Lean in Nix.

### Steps to Reproduce

This has to be done on macOS
1. Clone revision `b4caee80a3dfc5c9619d88b16c40cc3db90da4e2`
2.Execute `nix build .#packages.x86_64-darwin.libInit_shared` (this takes a while since the error occurs at link stage)
3.Error shows up

**Expected behavior:**

Lean should build

**Actual behavior:**

Fails to build due to the above error

### Versions

Revision `b4caee80a3dfc5c9619d88b16c40cc3db90da4e2`

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

### PR DESCRIPTION
**Pull request by lenianiva**
Closes #3810

### CODE REVIEW

**Comment by vleni** [comment]
WIP: Currently there's a missing symbol error:
```sh
$ nix build .#packages.x86_64-darwin.libInit_shared
```
gives
```
ld: warning: directory not found for option '-L/nix/store/v2y691apvl5y0crkwlzvpfkplvh09qx2-lean-bin-tools/lib/lean'
ld: warning: ignoring file /nix/store/m5mkq2bbq66z4xxyzprr39nlx80dbl0a-Lean-lib/libLean.a, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x7>ld: warning: ignoring file /nix/store/n8pzs12zzpd55h7zff6xqss0nv0dckmk-Init-lib/libInit.a, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x7>Undefined symbols for architecture x86_64:
  "_initialize_Init", referenced from:
      _lean_initialize in libleancpp.a(init.cpp.o)
      lean::initializer::initializer() in libleancpp.a(init.cpp.o)
      lean::initializer::initializer() in libleancpp.a(init.cpp.o)
  "_initialize_Lean", referenced from:
      _lean_initialize in libleancpp.a(init.cpp.o)
      lean::initializer::initializer() in libleancpp.a(init.cpp.o)
      lean::initializer::initializer() in libleancpp.a(init.cpp.o)
  "_initialize_Lean_Compiler_IR_EmitLLVM", referenced from:
      _lean_main in libleancpp.a(shell.cpp.o)
  "_lean_add_projection_info", referenced from:
      lean::save_projection_info(lean::environment const&, lean::name const&, lean::name const&, unsigned int, unsigned int, bool) in libleancpp.a(projection.cpp.o)
  "_lean_add_protected", referenced from:
      lean::add_protected(lean::environment const&, lean::name const&) in libleancpp.a(protected.cpp.o)
  "_lean_add_specialization_info", referenced from:
      lean::update_spec_info(lean::environment const&, lean::list_ref<lean::pair_ref<lean::name, lean::expr>> const&) in libleancpp.a(specialize.cpp.o)
  "_lean_array_to_list", referenced from:
      _lean_array_data in libleanrt_initial-exec.a(object.cpp.o)
...skipping...
      lean::recursor_val::is_k() const in libleancpp.a(declaration.cpp.o)
  "_lean_reducibility_hints_get_height", referenced from:
      lean::reducibility_hints::get_height() const in libleancpp.a(declaration.cpp.o)
      lean::compare(lean::reducibility_hints const&, lean::reducibility_hints const&) in libleancpp.a(declaration.cpp.o)
      std::__1::__function::__func<lean::get_max_height(lean::environment const&, lean::expr const&)::$_1, std::__1::allocator<lean::get_max_height(lean::environment const&, lean::expr const&)::$_1>, bool (l>  "_lean_register_option", referenced from:
      lean::register_option(lean::name const&, lean::name const&, lean::data_value_kind, char const*, char const*) in libleancpp.a(option_declarations.cpp.o)
  "_lean_run_frontend", referenced from:
      lean::run_new_frontend(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, lean::options const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::>  "_lean_server_watchdog_main", referenced from:
      lean::run_server_watchdog(lean::buffer<lean::string_ref, 16ul> const&) in libleancpp.a(shell.cpp.o)
  "_lean_server_worker_main", referenced from:
      lean::run_server_worker(lean::options const&) in libleancpp.a(shell.cpp.o)
      _lean_main in libleancpp.a(shell.cpp.o)
  "_lean_set_reducibility_status", referenced from:
      lean::set_reducible(lean::environment const&, lean::name const&, lean::reducible_status, bool) in libleancpp.a(reducible.cpp.o)
  "_lean_stream_of_handle", referenced from:
      lean::initialize_io() in libleanrt_initial-exec.a(io.cpp.o)
  "_lean_write_module", referenced from:
      lean::write_module(lean::environment const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) in libleancpp.a(module.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by jetjinser** [comment]
Any progress? I'm in the same predicament and I really want it to work on my x86_64-darwin. Is there anything I can do to help?

I finally found that missing symbol `_lean_initialize` is in leancpp, and this dependency is not in Leanc.executable.

Frankly speaking, I am a bit confused because I don't understand how lean is built 😔.

**Comment by nomeata** [comment]
From my side the status is that I'm rooting for you here, and fixes can be merged, but usage of nix beyond what we need for the CI setup need to be taken care of by those that use it. Hope you understand.

**Comment by lenianiva** [comment]
> Any progress? I'm in the same predicament and I really want it to work on my x86_64-darwin. Is there anything I can do to help?
> 
> I finally found that missing symbol `_lean_initialize` is in leancpp, and this dependency is not in Leanc.executable.
> 
> Frankly speaking, I am a bit confused because I don't understand how lean is built 😔.

can you poke around the build targets of `nix/bootstrap.nix` a bit to see what is going on? When I tried to fix it I looked at the symbol table of `libInit_shared` and `leanc` etc. to see if they have the symbols required, but I don't know why it still doesn't compile.

**Comment by lenianiva** [comment]
I'll periodically merge `master` into this branch and see if anything changes, but I have spent >8h on this and I could not make more progress

**Comment by lenianiva** [comment]
> Any progress? I'm in the same predicament and I really want it to work on my x86_64-darwin. Is there anything I can do to help?
> 
> I finally found that missing symbol `_lean_initialize` is in leancpp, and this dependency is not in Leanc.executable.
> 
> Frankly speaking, I am a bit confused because I don't understand how lean is built 😔.

How is `leancpp` not a dependency of `Leanc.executable`? The linker flags of `leanshared` has `-lleancpp`, and `leanshared` is a dependency of `leanc`

**Comment by lenianiva** [comment]
> Any progress? I'm in the same predicament and I really want it to work on my x86_64-darwin. Is there anything I can do to help?
> 
> I finally found that missing symbol `_lean_initialize` is in leancpp, and this dependency is not in Leanc.executable.
> 
> Frankly speaking, I am a bit confused because I don't understand how lean is built 😔.

I fixed it!

**Comment by jetjinser** [comment]
> > Any progress? I'm in the same predicament and I really want it to work on my x86_64-darwin. Is there anything I can do to help?
> > I finally found that missing symbol `_lean_initialize` is in leancpp, and this dependency is not in Leanc.executable.
> > Frankly speaking, I am a bit confused because I don't understand how lean is built 😔.
> 
> I fixed it!

marvelous! 😍

**Comment by nomeata** [comment]
@jetjinser, did you have a chance to review and test the fix?

**Comment by jetjinser** [comment]
> @jetjinser, did you have a chance to review and test the fix?

Yes! It actually built **successfully** on my x86_64-darwin machine!
It looks like nix specifies the correct lib on darwin.

File: nix/bootstrap.nix [hunk 1]
```diff
@@ -125,7 +125,7 @@ rec {
       leanshared = runCommand "leanshared" { buildInputs = [ stdenv.cc ]; libName = "libleanshared${stdenv.hostPlatform.extensions.sharedLibrary}"; } ''
         mkdir $out
         LEAN_CC=${stdenv.cc}/bin/cc ${lean-bin-tools-unwrapped}/bin/leanc -shared ${lib.optionalString stdenv.isLinux "-Wl,-Bsymbolic"} \
-          ${if stdenv.isDarwin then "-Wl,-force_load,${Init.staticLib}/libInit.a -Wl,-force_load,${Lean.staticLib}/libStd.a -Wl,-force_load,${Lean.staticLib}/libLean.a -Wl,-force_load,${leancpp}/lib/lean/libleancpp.a ${leancpp}/lib/libleanrt_initial-exec.a -lc++"
+          ${if stdenv.isDarwin then "-Wl,-force_load,${Init.staticLib}/libInit.a -Wl,-force_load,${Lean.staticLib}/libLean.a -Wl,-force_load,${leancpp}/lib/lean/libleancpp.a ${leancpp}/lib/libleanrt_initial-exec.a -lc++"
```
Comment (tydeu) [inline nix/bootstrap.nix:128]: I think `libStd` was accidently lost here.

Comment (lenianiva) [reply nix/bootstrap.nix:128]: but somehow it still builds and runs despite libStd not being there? Anyways I'll add it back in

Comment (lenianiva) [reply nix/bootstrap.nix:128]: its added back in https://github.com/leanprover/lean4/pull/4587#event-13338702834

**Pull request merged**